actions/channel_actions.ts

```diff
@@ -179,3 +195,120 @@ export function muteChannel(userId: UserProfile['id'], channelId: Channel['id'])
         mark_unread: NotificationLevels.MENTION,
     });
 }
+
+/**
+ * Fetches channels and channel members with graphql and then dispatches the result to redux store.
+ * @param teamId If team id is provided, only channels and channel members in that team will be fetched. Otherwise, all channels and all channel members will be fetched.
+ */
+export function fetchChannelsAndMembers(teamId: Team['id'] = ''): ActionFunc<{channels: ServerChannel[]; channelMembers: ChannelMembership[]}> {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const currentUserId = getCurrentUserId(state);
+
+        let channelsResponse: ChannelsQueryResponseType['data']['channels'] = [];
+        let channelMembersResponse: ChannelMembersQueryResponseType['data']['channelMembers'] = [];
+        let roles: Role[] = [];
+
+        try {
+            const {data: channelsAndChannelMembers, errors} = await Client4.fetchWithGraphQL<ChannelsAndChannelMembersQueryResponseType>(getChannelsAndChannelMembersQueryString(teamId));
+
+            if (errors || !channelsAndChannelMembers) {
+                throw new Error('Error returned in fetching channels and channel members of first page');
+            }
+
+            // First page of channels and channel members
+            channelsResponse = [...channelsAndChannelMembers.channels];
+            channelMembersResponse = [...channelsAndChannelMembers.channelMembers];
+
+            if (channelsResponse && channelsResponse.length === CHANNELS_MAX_PER_PAGE) {
+                let channelsPerPageDataLength;
+                let channelsCursor = channelsResponse[CHANNELS_MAX_PER_PAGE - 1].cursor;
+
+                do {
+                    const {data: channelsPerPageResponse, errors} = await Client4.fetchWithGraphQL<ChannelsQueryResponseType>(getChannelsQueryString(channelsCursor, teamId)); // eslint-disable-line no-await-in-loop
```

Comment:
This isn't something that needs to be changed, but I noticed that we're paging through channels here and then through channel members below because we're getting the user's channels and channel members separately in the query.

If that ever causes issues because we're making too many requests or something, I wonder if we could change it so that instead of requesting the channels and members as two separate arrays, we instead request a single array of their channel members with the channels attached to those. That way, there should only be one cursor and we know that we should always have the matching channels and members

Reply:
Yes, let me change it to single loop

Reply:
Just curious though i tested this one out locally for both admins and members, the number of channels and members were same. Under what circumstances can they be different though 🤔 if it all they were to be different.

Reply:
@larkox @hmhealey made it into single loop